### PR TITLE
Don't load all ancestors when truncating

### PR DESCRIPTION
**Pull request by benjeffery**

### CODE REVIEW

**Comment by benjeffery** [comment]
This was way too slow at around 8hours for the smallest chrom in my dataset. Have just pushed a change which I'm now testing where only changed ancestors are touched.

**Comment by benjeffery** [comment]
@mergifyio rebase

**Comment by benjeffery** [comment]
Fixed up, should be good to merge.

File: tests/test_inference.py [hunk 1]
```diff
@@ -568,6 +568,72 @@ def verify_data_round_trip(
...
+            sequence_length,
+            site_times,
+            individual_times,
+            ancestral_alleles,
+        )
+        with tempfile.TemporaryDirectory() as d:
+            tsinfer.generate_ancestors(sample_data, path=d + "ancestors.tsi")
+            ancestors = tsinfer.AncestorData.load(d + "ancestors.tsi")
+            time = np.sort(ancestors.ancestors_time[:])
+            if (
```
Comment (jeromekelleher) [inline tests/test_inference.py:600]: Maybe put the comment on a different line? Seems like unnecessary line breakage here

File: tests/test_inference.py [hunk 2]
```diff
@@ -568,6 +568,72 @@ def verify_data_round_trip(
...
+                midpoint = np.median(time)
+                params = [
+                    (lower_bound, upper_bound, 0.1),
+                    (lower_bound, upper_bound, 1),
+                    (midpoint, midpoint + (midpoint / 2), 1),
+                ]
+            else:
+                params = [(0.4, 0.6, 1), (0, 1, 10)]
+            for param in params:
+                truncated_ancestors = ancestors.truncate_ancestors(
```
Comment (jeromekelleher) [inline tests/test_inference.py:614]: will ``ancestors.truncate_ancestors(*param, buffer_length=2)``  work here ?

File: tsinfer/formats.py [hunk 3]
```diff
@@ -3066,7 +3064,23 @@ def truncate_ancestors(
             raise ValueError("No ancestors in time bound")
         max_length = length_multiplier * np.max(self.ancestors_length[:][anc_in_bound])
 
-        for anc in self.ancestors():
+        truncated = self.copy(**kwargs)
+
+        # Create a buffer of 1000 ancestors with their indexes
```
Comment (jeromekelleher) [inline tsinfer/formats.py:3069]:
```suggestion
        # Create a buffer of buffer_length ancestors with their indexes
```

File: tsinfer/formats.py [hunk 4]
```diff
@@ -3088,21 +3102,57 @@ def truncate_ancestors(
...
                     ] = anc.full_haplotype[insert_pos_start:insert_pos_end]
-                    # TODO - record truncation in ancestors' metadata when supported
-        truncated.ancestors_start[:] = start
-        truncated.ancestors_end[:] = end
-        truncated.ancestors_time[:] = time
-        truncated.ancestors_focal_sites[:] = focal_sites
-        truncated.ancestors_full_haplotype[:] = haplotypes
-        truncated.ancestors_full_haplotype_mask[:] = haplotypes == tskit.MISSING_DATA
+                    buffer_pos += 1
+                    if buffer_pos == buffer_length:
```
Comment (jeromekelleher) [inline tsinfer/formats.py:3114]: A local function would help here,

```python
def flush_buffer(length):
       truncated.ancestors_start.set_orthogonal_selection(
              index_buffer[:length], start_buffer[:length]
         )
        # etc
```

Comment (jeromekelleher) [reply tsinfer/formats.py:3114]: I don't follow how this is working, so a few comments on how ``flush_buffer`` works and would be helpful here.

**Review by jeromekelleher — APPROVED** [review]
LGTM, a few suggestions. Happy to merge when addressed so pre-approving

**Comment by benjeffery** [comment]
Fixed up, merging.

**Pull request merged**